### PR TITLE
Add interpreter for ReverseOp

### LINKED ISSUE #985: Add interpreter for reverse
Following the [testing guidelines](https://github.com/openxla/stablehlo/blob/main/docs/interpreter_checklist.md)

### PR DESCRIPTION
**Pull request by ghpvnist**
closes #985

### CODE REVIEW

**Review by subhankarshah — APPROVED** [review]
LGTM, neat.

File: stablehlo/reference/Ops.cpp [hunk 1]
```diff
@@ -178,6 +178,22 @@ Tensor eval(ReshapeOp op, const Tensor &operand) {
...
+       resultIt != result.index_end(); ++resultIt, ++operandIt) {
+    auto operandIdx = (*operandIt).vec();
+    for (auto dim : dimensions) {
+      operandIdx[dim] = -(operandIdx[dim] - (shape[dim] - 1));
+    }
+    result.set(*resultIt, operand.get(operandIdx));
+  }
+  return result;
+}
+
```
Comment (sdasgup3) [inline stablehlo/reference/Ops.cpp:196]: Suggestion:
1. Avoid loop in collecting the dimensions in a `SmallVector`.
2. As per the spec, given an index of result we can derive the index of the operand. With that said, 
  -  we can avoid the second iterator. 
  -  make the implementation follow more closely to the spec
3. Prefer `SmallVector` over `std::vector`

```
Tensor eval(ReverseOp op, const Tensor &operand) {
  Tensor result(op.getType());
  auto shape = op.getType().getShape();
  auto dimensions = op.getDimensions().getValues<int64_t>();

  for (auto resultIt = result.index_begin(); resultIt != result.index_end();
       ++resultIt) {
    SmallVector<int64_t> operandIdx(*resultIt);
    for (auto dim : dimensions) {
      operandIdx[dim] = (shape[dim] - 1) - operandIdx[dim];
    }
    result.set(*resultIt, operand.get(operandIdx));
  }
  return result;
}

```

**Comment by sdasgup3** [comment]
Other than the suggested change, the PR LGTM.

**Review by sdasgup3 — APPROVED** [review]

File: docs/spec.md [hunk 2]
```diff
@@ -4143,16 +4143,6 @@ and produces a `result` tensor. More formally,
 #### Examples
 
 ```mlir
```
Comment (sdasgup3) [inline docs/spec.md:4145]: Link to test file `interpreter_reverse.mlir` is missing

File: stablehlo/reference/Ops.cpp [hunk 3]
```diff
@@ -178,6 +178,20 @@ Tensor eval(ReshapeOp op, const Tensor &operand) {
   return result;
 }
 
+Tensor eval(ReverseOp op, const Tensor &operand) {
```
Comment (burmako) [inline stablehlo/reference/Ops.cpp:181]: This will need to be rebased on top of #1007.

File: stablehlo/reference/Ops.cpp [hunk 4]
```diff
@@ -178,6 +178,20 @@ Tensor eval(ReshapeOp op, const Tensor &operand) {
...
+  Tensor result(op.getType());
+  auto shape = op.getType().getShape();
```
Comment (burmako) [inline stablehlo/reference/Ops.cpp:183]: `resultShape` to make it clear what shape exactly this is referring to?

File: stablehlo/tests/ops_stablehlo.mlir [hunk 5]
```diff
@@ -3046,7 +3056,7 @@ func.func @reverse_duplicate_dimensions(%operand: tensor<3x2xi32>) -> tensor<3x2
 
 // -----
 
-func.func @reverse_invalid_dimensions_unranked(%operand: tensor<*xi32>) -> tensor<*xi32> {
+func.func @reverse_C3(%operand: tensor<*xi32>) -> tensor<*xi32> {
```
Comment (burmako) [inline stablehlo/tests/ops_stablehlo.mlir:3059]: How about calling this `@reverse_c3` to align with the snake case naming convention used in the file?

File: stablehlo/tests/interpret_reverse.mlir [hunk 6]
```diff
@@ -0,0 +1,53 @@
+// RUN: stablehlo-interpreter --interpret -split-input-file %s | FileCheck %s
```
Comment (burmako) [inline stablehlo/tests/interpret_reverse.mlir:1]: Do we need all the three tests here or just one is sufficient, similarly to `ops_stablehlo.mlir` which has just one positive test?

File: stablehlo/dialect/TypeInference.cpp [hunk 7]
```diff
@@ -3728,15 +3728,18 @@ LogicalResult verifyReverseOp(std::optional<Location> location, Value operand,
...
     return emitOptionalError(location,
                              "dimensions should be unique. Got: ", dims);
   auto operandTy = operand.getType().dyn_cast<RankedTensorType>();
   for (int64_t dim : uniqueDims) {
+    // ReverseOp_C3
     if (dim < 0)
       return emitOptionalError(
           location,
           "all dimensions should be non-negative. Got dimension: ", dim, ".");
+    // ReverseOp_C3
```
Comment (burmako) [inline stablehlo/dialect/TypeInference.cpp:3742]: Let's perhaps drop this line, given that we're already under `ReverseOp_C3`?

**Review by burmako — CHANGES_REQUESTED** [review]

File: stablehlo/tests/ops_stablehlo.mlir (hunk 5 above)
Comment (ghpvnist) [reply stablehlo/tests/ops_stablehlo.mlir:3059]: I followed the guideline in the [checklist](https://github.com/openxla/stablehlo/blob/main/docs/reference_checklist.md). Perhaps tests (ops_stablehlo.mlir and interpret_foo.mlir), comments, test names should all be snake lowercase without appending "op"?
```
// slice_c1

func.func @slice_c2
```

Comment (burmako) [reply stablehlo/tests/ops_stablehlo.mlir:3059]: Sounds reasonable to me. cc @sdasgup3

Comment (sdasgup3) [reply stablehlo/tests/ops_stablehlo.mlir:3059]: LGTM

**Review by burmako — APPROVED** [review]

**Pull request merged**